portable/GCC/ARM_CM3/port.c

```diff
@@ -293,6 +289,18 @@ BaseType_t xPortStartScheduler( void )
         /* Use the same mask on the maximum system call priority. */
         ucMaxSysCallPriority = configMAX_SYSCALL_INTERRUPT_PRIORITY & ucMaxPriorityValue;
 
+        /* Check that the maximum system call priority is nonzero after
+         * accounting for the number of priority bits supported by the
+         * hardware. A priority of 0 is invalid because setting the BASEPRI
+         * register to 0 unmasks all interrupts, and interrupts with priority 0
+         * cannot be masked using BASEPRI.
+         * See https://www.FreeRTOS.org/RTOS-Cortex-M3-M4.html */
+        configASSERT( ucMaxSysCallPriority );
+
+        /* Check that the interrupt priority used for PendSV and SysTick is the
+         * lowest supported by the hardware. */
+        configASSERT( (configKERNEL_INTERRUPT_PRIORITY & ucMaxPriorityValue) == ucMaxPriorityValue );
```

Comment:
Even though it is inline with out recommendation, the second assert does not seem backward compatible and it may break some existing application.  Can we avoid adding this assert?

Also, would you please make the same change in [IAR ports ](https://github.com/FreeRTOS/FreeRTOS-Kernel/tree/main/portable/IAR)as well?

Reply:
Such applications are already broken though, right? I could say the same about the max syscall assert. My tests appeared to work for a while until I dialed optimizations up. The PendSV handler uses the exception return value and stores it in a task's stack. This doesn't make sense if PendSV interrupted another interrupt. Switching to a different task's exception return value will then cause another exception due to the mismatch of returning to thread mode while another exception is already pending. @jefftenney thoughts?

I will take a stab at the IAR version.

Reply:
I agree that PendSV must use the lowest priority.  That is a requirement of these ports.  But this discussion does bring up a concern about backward compatibility.

If this new assertion fails for an existing application, the likely effect is developer anxiety. They might wonder why do I have to reduce the kernel interrupt priority?  What am I going to break if I do that?  The reality is that by reducing the kernel interrupt priority they would be fixing their application, but they won't necessarily know that.

One alternative is to modify these ports to configure PendSV for the lowest priority regardless of how `configKERNEL_INTERRUPT_PRIORITY` is set.  The ARMv8-M ports do this.

Reply:
Makes sense to me. I've updated the PR to do this, and applied the change to all the different toolchain variations of the ports. Some of them had default settings of `configKERNEL_INTERRUPT_PRIORITY` and comments suggesting this is for backward compatibility, so I've left those alone, but others no longer mention this macro at all.

Reply:
Some of the ports also already had an assert on `configKERNEL_INTERRUPT_PRIORITY & ucMaxPriorityValue`, of the form you suggested originally, but I've now removed those as they are no longer necessary.

Reply:
Good points and thank you for your changes. 

>Some of them had default settings of configKERNEL_INTERRUPT_PRIORITY and comments suggesting this is for backward compatibility, so I've left those alone, but others no longer mention this macro at all.

I can see these default definition of configKERNEL_INTERRUPT_PRIORITY in the following ports - 
* portable/GCC/ARM_CM3/port.c
* portable/IAR/ARM_CM0/port.c
* portable/IAR/ARM_CM3/port.c
* portable/RVDS/ARM_CM3/port.c

Given that this macro is no longer used, I think these can se safely removed.

The same change is needed in this port too - portable/Tasking/ARM_CM4F/port.c

Once again, thank you for your contribution.

Reply:
If we remove use of configKERNEL_INTERRUPT_PRIORITY we should introduce a #warning "Version x.y.z of the FreeRTOS kernel removed the configKERNEL_INTERRUPT_PRIORITY constant as it should always be set to the lowest priority number." - or words to that effect.

As an aside, there would also be a lot of documentation to update, separate from this PR.

Reply:
Sure. Can you add the following to each of the port modified port file - 

```c
#ifdef configKERNEL_INTERRUPT_PRIORITY
    #warning  configKERNEL_INTERRUPT_PRIORITY is no longer used as it should always be set to the lowest priority number. Please remove it from your FreeRTOSConfig.h.
 #endif
```

@ChristosZosi - Let me know if you are okay making these change or you'd like us to do those for you.

Reply:
I've made the change to the Tasking/ARM_CM4F port as well. Its `xPortStartScheduler` doesn't contain the additional asserts and the code that determines the number of bits the hardware support, so I've omitted the max syscall priority masking change there as well.

As for the warning if `configKERNEL_INTERRUPT_PRIORITY` is defined, I defer to the FreeRTOS team to handle those. I'm not aware of how all the different toolchains do custom diagnostic messages.